### ISSUE TITLE
Advanced Search for Ministry Users

### ISSUE THREAD

**Issue by m-prodan** [bcgov/foi-flow#1654]
* As a Ministry user
* I want to be able to search for requests
* so that I can refer to past requests and decisions made

**Assumptions & Scope**	
This story covers the need for Ministry users to have an advanced search on their queue

What is IN scope?
This will cover the Ministry queue and functionality for Ministry users that cover a single Ministry, and has a special AC for the economy sector. There may be future enhancements or stories to address different Ministry coordinators who have unique needs similar to economy sector.

What is NOT in scope?
IAO Advanced Search - this is covered in user story #1213

**Acceptance Criteria**	

**Scenario 1: Advanced Search Layout**
* GIVEN a ministry coordinator is on their queue
* AND they are assigned to a single Ministry
* WHEN the user clicks the 'Advanced Search' toggle
* THEN the advanced search view will open
* AND there will be a search field, search buttons (AXIS request #, Raw request #, Applicant Name, Assignee Name, Request Description, Subject Code), Request State checkboxes (Call for Record, Records Review, Ministry Sign off, Closed), Request Status Checkboxes( All Active, Overdue, On Time), Request Type checkboxes (General, Personal) Search date range (From, To), Select Public Body (Locked Field), and a button to 'Apply Search' (primary action), and a button to 'Clear All Filters' (secondary action).

**Scenario 2: Single Ministry User Role - Select Public Body**
* GIVEN a Ministry coordinator has clicked on the advanced search
* AND the Ministry coordinator is assigned to a single Ministry
* WHEN they click on the "Select Public Body" dropdown
* THEN they will be unable to activate the drop down
* AND the selection will be locked to the Ministry the user is assigned to

**Scenario 3: Multi-Ministry User Role - Select Public Body**
* GIVEN a Ministry coordinator has clicked on the advanced search
* AND the Ministry coordinator is assigned to multiple Ministries
* WHEN they click on the "Select Public Body" dropdown
* THEN the drop down will be active
* AND the selection's available will only be to the Ministries the coordinator is assigned to.

Scenario 4: Open Date Calendar
* GIVEN a Ministry Coordinator is on the advanced search view
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/874445058
* WHEN the analyst clicks the calendar icon next to the From Date or To Date
* THEN the calender view will open
* AND the analyst cannot select a 'To Date' before the 'From Date'

Scenario 5: Edit Date 2
* GIVEN the Coordinator has opened the calendar for the 'From Date' or 'To Range (
* WHEN they click on a date in the calendar (standard material UI function)
* THEN that date will be selected

Scenario 6: Edit Date Range Start Date 3
* GIVEN the coordinator has selected a date
* WHEN they click 'anywhere on the screen outside the calendar
* THEN the calendar will close
* AND the selected date will be displayed
* AND the selected date will be the new 'From Date' or 'To Date' diplayed
* AND those dates will be filters for the search

Scenario 7: Cannot select to 'To Date' before the 'From Date'
* GIVEN the coordinator has opened the calendar
https://jacklyn808742.invisionapp.com/console/share/XU2L77IGEA/631957829
* WHEN they click a 'To Date' before the 'From Date'
* THEN date will not be able to be selected
* AND a message will appear that you cannot select a 'To Date' before the 'From Date'

Scenario 8: Select a Request State, Request Status, Request Type
* GIVEN a Ministry Coordinator is on the advanced search view
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/874445058
* WHEN they click a box for Request State, Request Status, Request Type
* THEN the search will be filtered by that selection
* AND only one box can be checked

Scenario 9: Select a pre-filter button (e.g. AXIS Request #)
* GIVEN a Ministry Coordinator is on the advanced search view
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/874445058
* WHEN they click a button (e.g. AXIS Request #)
* THEN the search will be filtered by that selection
* AND only one button can be checked

Scenario 10: 'Clear All Filters
* GIVEN a Ministry Coordinator is on the advanced search view
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/874445058
* AND selections have been made
* WHEN they click the 'Clear All Filter' button
* THEN all search filters will be removed

Scenario 11: Apply Search
* GIVEN a Ministry Coordinator is on the advanced search view
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/874445058
* AND selections have been made
* WHEN they click the 'Apply Search' button
* THEN requests will be returned based on the selected search criteria and listed below the Advanced Search in the same order with the same columns at they do on the queue

Scenario 14: Return to queue
* GIVEN a Ministry Coordinator is on the advanced search view
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/874445058
* WHEN the user clicks the 'Your FOI Request' toggle
* THEN the ministry coordinator will be returned to their queue
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/874445057

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by m-prodan** [comment]
Separate story for advanced search for Ministry role as discussed earlier today with @abin-aot @richard-aot @nkan-aot @lmullane @JHarrietha-AOT 

Jackie - I believe we may need some updated designs based on our discussion. I will likely need to revise part of AC1 - does AC2 and 3 make sense based on our discussions? 

Thanks!

**Comment by nkan-aot** [comment]
Dependent on #1634

**Comment by JHarrietha-AOT** [comment]
@abin-aot @richard-aot Do you feel like you need updated designs? Let me know if you do.

As discussed yesterday Unopened would be disabled (light grey). And if it is a Ministry where they are a single ministry than the Public Body dropdown will be disabled. If they are a multiple Ministry User then they will have the public body dropdown with ONLY there respective Ministries

**Comment by richard-aot** [comment]
Thanks for clarify this, @JHarrietha-AOT . Don't need an update, I can work with the current design.

**Comment by m-prodan** [comment]
Updated AC to include all the typical functionality noted in #1213 - let me know if there are any questions.

**Comment by richard-aot** [comment]
Hi team, this US is ready for testing. We made the following updates:
1. Add advanced search for ministry users
2. Change selected checkbox color to green
3. Fix the issue that unselect the last public body, the list jumps to the top
4. Format assignedto name on backend - fullname/group name
5. Sorting assignedto name will apply to both fullname and group name
6. Fixed sorting issue on LDD  & Days left - IAO queue
7. Fixed wrong received date issue - IAO queue

**Comment by JHarrietha-AOT** [comment]
Great work on these, @richard-aot just wanted to highlight here in the comment that we have Ticket #1688 upcoming which will address the fact that Scenario's 9 & 10 do not pass testing. They fail because "Request Description" is still selected if a users clicks "Clear All"
